File: code/ch12/12.3.1.nimGame.py
class Solution:
    def canWinNim(self, n: int) -> bool:
        mem = [True] * (n + 1)
        for i in range(4, n + 1):
            if not (mem[i - 1] and mem[i - 2] and mem[i - 3]):
                mem[i] = True
            else:
                mem[i] = False
        return mem[n]

File: code/ch12/12.3.1.nimGame.cpp
//TLE
class Solution {
public:
    bool canWinNim(int n) {
	bool res = false;
        bool *mem = (bool *)malloc((n + 1) * sizeof(bool));
        memset(mem, true, (n + 1) * sizeof(bool));
        
        for (int i = 4; i < n + 1; i++) {
            if (!(mem[i - 1] && mem[i - 2] && mem[i - 3]))
                mem[i] = true;
            else
                mem[i] = false;
        }
        
	res = mem[n];
	free(mem);
        return res;
    }
};
